packages/cloudflare-workers-bindings-extension/src/add-binding.ts

```diff
@@ -0,0 +1,329 @@
+		public label: string,
+		public description?: string,
+		public detail?: string,
+		public iconPath?: Uri
+	) {}
+}
+
+export async function multiStepInput(context: ExtensionContext) {
+	const bindingTypes: BindingType[] = [
+		new BindingType(
+			"KV",
+			"kv_namespaces",
+			"Global, low-latency, key-value data storage",
+			Uri.file(context.asAbsolutePath("resources/icons/kv.svg"))
+		),
+		new BindingType(
+			"R2",
+			"r2_buckets",
+			"Object storage for all your data",
+			Uri.file(context.asAbsolutePath("resources/icons/r2.svg"))
+		),
+		new BindingType(
+			"D1",
+			"d1_databases",
+			"Serverless SQL databases",
+			Uri.file(context.asAbsolutePath("resources/icons/d1.svg"))
+		),
+	];
+
+	interface State {
+		title: string;
+		step: number;
+		totalSteps: number;
+		bindingType: BindingType;
+		name: string;
+		runtime: QuickPickItem;
+		id: string;
+	}
+
+	async function collectInputs() {
+		const state = {} as Partial<State>;
+		await MultiStepInput.run((input) => pickResourceGroup(input, state));
+		return state as State;
+	}
+
+	const title = "Add binding";
+
+	async function pickResourceGroup(
+		input: MultiStepInput,
+		state: Partial<State>
+	) {
+		const pick = await input.showQuickPick({
+			title,
+			step: 1,
+			totalSteps: 2,
+			placeholder: "Choose a binding type",
+			items: bindingTypes,
+			activeItem:
+				typeof state.bindingType !== "string" ? state.bindingType : undefined,
+		});
+		state.bindingType = pick as BindingType;
+		return (input: MultiStepInput) => inputName(input, state);
+	}
+
+	async function inputName(input: MultiStepInput, state: Partial<State>) {
+		let name = await input.showInputBox({
+			title,
+			step: 2,
+			totalSteps: 2,
+			value: state.name || "",
+			prompt: "Choose a binding name",
+			validate: validateNameIsUnique,
+			placeholder: `e.g. MY_BINDING`,
+		});
+		state.name = name;
+		return () => addToToml(state);
+	}
+
+	async function addToToml(state: Partial<State>) {
+		const configUri = await getConfigUri();
+		if (!configUri) {
+			return null;
+		}
+		const workspaceFolder = workspace.getWorkspaceFolder(configUri);
+
+		if (!workspaceFolder) {
+			return null;
+		}
+
+		const wrangler = importWrangler(workspaceFolder.uri.fsPath);
+
+		await workspace.openTextDocument(configUri).then((doc) => {
+			window.showTextDocument(doc);
+			try {
+				wrangler.experimental_patchConfig(configUri.path, {
+					[state.bindingType?.description!]: [{ binding: state.name! }],
+				});
+				window.showInformationMessage(`Created binding '${state.name}'`);
+			} catch {
+				window.showErrorMessage(
+					`Unable to directly add binding to config file. A snippet has been copied to clipboard - please paste this into your config file.`
+				);
+
+				const patch = `[[${state.bindingType?.description!}]]
+binding = "${state.name}"
+`;
+
+				env.clipboard.writeText(patch);
+			}
+		});
+	}
+
+	async function validateNameIsUnique(name: string) {
+		// TODO: actually validate uniqueness
```

Comment:
Could you make an issue to track this?

Reply:
[DEVX-1547](https://jira.cfdata.org/browse/DEVX-1547)